File: python/task3c/S3_PIPE_06.py
# S3_PIPE_06

def pipe_safe(*fns):
    def wrapper(x):
        val = x
        try:
            for f in fns:
                val = f(val)
            return {"ok": True, "value": val}
        except Exception as e:
            return {"ok": False, "error": str(e)}
    return wrapper

def safe_div(x):
    return 10 / x

pipeline = pipe_safe(lambda x: x - 2, safe_div)

print(pipeline(4)) # (4-2)=2 -> 10/2=5 -> OK
print(pipeline(2)) # (2-2)=0 -> div error -> Fail

File: js/task3/S3_PIPE_06.js
// S3_PIPE_06

function pipeSafe(...fns) {
    return (input) => {
        try {
            let result = input;
            for (let fn of fns) {
                result = fn(result);
            }
            return { ok: true, value: result };
        } catch (e) {
            return { ok: false, error: e.message };
        }
    };
}

const safeDiv = x => {
    if (x === 0) throw new Error("Zero division");
    return 10 / x;
};

const pipeline = pipeSafe(x => x - 2, safeDiv);

console.log(pipeline(4)); // (4-2)=2 -> 10/2=5 -> ok
console.log(pipeline(2)); // (2-2)=0 -> Error -> not ok